### PR TITLE
basic vault integration

### PR DESCRIPTION
**Pull request by javib51**
@bkono what do you think?

### CODE REVIEW

**Comment by asim** [comment]
I'm looking at the code and on first glance it looks good. I'm going to merge and then let @bkono come back to review when he has time. Otherwise thanks for this. Its a great contribution.

**Comment by bkono** [comment]
@javib51 Sorry it took so long to come back and review properly. Took a first glance through awhile back and noted a few questions, then got buried. 

Can you do a gist of how your use cases are used with this implementation? There's some conventions that I think we can work through to make it a smoother experience, but also (at first glance) I don't think it can directly meet any of my current usages of vault. Before suggesting too many changes, I'd love to see how you're using it now, to make sure I don't suggest something that'll conflict.

File: source/vault/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Vault Source
+
+The vault source reads config from different secret engines in a Vault server. For example:
+```
+kv: secret/data/<my/secret>
+database credentials: database/creds/<my-db-role>
```
Comment (bkono) [inline source/vault/README.md]: An example from my earlier comment, I'd love to see a `WithDatabaseCreds` option that can take multiple databases, and will extract username/password, placing them under `{db: {username: ..., password: ...}}` for a single, or the first element of the path when given multiple. 
Example: 
`WithDatabaseCreds("legacydb/creds/myrole", "paymentsdb/creds/myrole")`

And get a result of: 
```
{
	"legacydb": {
		"username": "myrole-username-generated-abc123",
		"password": "uuid-stuff"
	},
	"paymentsdb": {
		"username": "myrole-username-generated-abc123",
		"password": "uuid-stuff"
	}
}
```

File: source/vault/options.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	return func(o *source.Options) {
+		if o.Context == nil {
+			o.Context = context.Background()
+		}
+		o.Context = context.WithValue(o.Context, addressKey{}, a)
+	}
+}
+
+// WithResourcePath sets the resource that will be access
+func WithResourcePath(p string) source.Option {
```
Comment (bkono) [inline source/vault/options.go]: Nearly all of my applications need multiple paths. Swapping this to variadic args and storing a slice would work well for those cases.

**Comment by javib51** [comment]
> @javib51 Sorry it took so long to come back and review properly. Took a first glance through awhile back and noted a few questions, then got buried.
> 
> Can you do a gist of how your use cases are used with this implementation? There's some conventions that I think we can work through to make it a smoother experience, but also (at first glance) I don't think it can directly meet any of my current usages of vault. Before suggesting too many changes, I'd love to see how you're using it now, to make sure I don't suggest something that'll conflict.

Sure!
I have multiple examples, but this is the most descriptive one. [here](https://gist.github.com/javib51/9a8b38a20600cfecccffa8b9940d482a)

Anyway, I'm starting to use this library. If you feel that there is a better way to do the interactions, we can change it. The idea is to make it as smooth as possible.

File: source/vault/README.md (hunk 1 above)
Comment (javib51) [reply source/vault/README.md]: I think that its a good idea, this and next comment.

**Comment by javib51** [comment]
Maybe we can create a new PR and work there in the changes. @bkono what do you think?

**Pull request merged**